router/src/batcher.rs

```diff
@@ -1,77 +1,77 @@
 /// Batching and inference logic
-use crate::{Db, Entry};
+use crate::Entry;
 use crate::{ErrorResponse, GenerateRequest};
 use axum::http::StatusCode;
 use axum::Json;
 use std::future::Future;
-use std::sync::Arc;
+use nohash_hasher::IntMap;
 use text_generation_client::{Batch, ClientError, GeneratedText, ShardedClient};
 use thiserror::Error;
-use tokio::sync::{oneshot, Notify};
+use tokio::sync::oneshot;
+use tokio::sync::mpsc::{channel, Permit, Sender};
+use tokio::sync::mpsc::error::TrySendError;
 use tokio::time::Instant;
 use tracing::instrument;
+use crate::queue::Queue;
 
 /// Batcher
 #[derive(Clone)]
 pub struct Batcher {
-    /// Request database
-    db: Db,
-    /// Shared state
-    shared: Arc<Shared>,
+    /// Request queue
+    sender: Sender<Entry>,
 }
 
-/// Batcher shared state
-struct Shared {
-    /// Batching background Tokio task notifier
-    batching_task: Notify,
-}
 
 impl Batcher {
     pub(crate) fn new(
         client: ShardedClient,
         max_batch_size: usize,
         max_waiting_tokens: usize,
+        queue_size: usize,
     ) -> Self {
-        // Batcher shared state
-        let db = Db::new();
-        let shared = Arc::new(Shared {
-            batching_task: Notify::new(),
-        });
+        // Set up queue
+        let (sender, receiver) = channel(queue_size);
 
         // Spawn batching background task that contains all the inference logic
         tokio::spawn(batching_task(
             client,
             max_batch_size,
             max_waiting_tokens,
-            db.clone(),
-            shared.clone(),
+            Queue::new(receiver),
         ));
 
-        Self { db, shared }
+        Self { sender }
+    }
+
+    /// Reserve a slot in the queue for sending a request
+    pub(crate) fn reserve_slot(&self) -> Result<RequestSender<'_>, TrySendError<()>> {
+        self.sender.try_reserve().map(|permit| RequestSender { permit })
     }
+}
+
+pub(crate) struct RequestSender<'a> {
```

Comment:
If reserve_slot is used inside infer, I don't think you need this struct.

Reply:
Yes that's right, and you don't need reserve at all then, can just call `send` directly on the channel. But the reason for doing it is explained in above comment.